### PR TITLE
Simplify `hir::PathSegment`

### PR DESCRIPTION
**Pull request by nnethercote**
r? @petrochenkov

### CODE REVIEW

**Comment by petrochenkov** [comment]
I think it's a good idea, since these IDs and resolutions are already actively used by lints and similar, it's better to make sure that they are always set.

File: compiler/rustc_hir/src/hir.rs [hunk 1]
```diff
@@ -202,13 +202,10 @@ impl Path<'_> {
...
-    // any path segments without these will not have save-analysis info and
-    // therefore will not have 'jump to def' in IDEs, but otherwise will not be
-    // affected. (In general, we don't bother to get the defs for synthesized
-    // segments, only for segments which have come from the AST).
-    pub hir_id: Option<HirId>,
-    pub res: Option<Res>,
+
+    pub hir_id: HirId,
+
+    pub res: Res,
```
Comment (petrochenkov) [inline compiler/rustc_hir/src/hir.rs:208]: Nit: weird space between fields.

File: compiler/rustc_hir/src/hir.rs [hunk 2]
```diff
@@ -226,12 +223,12 @@ pub struct PathSegment<'hir> {
 
 impl<'hir> PathSegment<'hir> {
     /// Converts an identifier to the corresponding segment.
-    pub fn from_ident(ident: Ident) -> PathSegment<'hir> {
-        PathSegment { ident, hir_id: None, res: None, infer_args: true, args: None }
+    pub fn from_ident(ident: Ident, hir_id: HirId, res: Res) -> PathSegment<'hir> {
```
Comment (petrochenkov) [inline compiler/rustc_hir/src/hir.rs:226]:
```suggestion
    pub fn new(ident: Ident, hir_id: HirId, res: Res) -> PathSegment<'hir> {
```
With these two new parameters it's no longer "from ident".

**Comment by Dylan-DPC** [comment]
failed in [rollup](https://github.com/rust-lang/rust/pull/101314#issuecomment-1235241191)

@bors r-

**Comment by nnethercote** [comment]
I fixed the rollup failure.

**Comment by nnethercote** [comment]
Weird regressions here. I did some downloads and diffs with Cachegrind. The `tt-muncher` and `html5ever-0.26.0` ones are due to changes in `parse_tt`, which this PR didn't touch, but which is hot and I think somewhat prone to perturbations due to minor changes. The `keccak` one is due to `process_obligations`, which is also hot and prone to perturbations. And I didn't expect the changes in this PR to have much effect, it really wasn't changing things much.

**Comment by nnethercote** [comment]
Some evidence [here](https://github.com/rust-lang/rust/pull/101464#issuecomment-1238064407) that these apparent regressions weren't real.

**Pull request merged**